### ISSUE TITLE
[TODO] [Day 9] EfficientNet에서 LayoutLMv3로 전환 및 최적화

### ISSUE THREAD

**Issue by nimnusmik** [AIBootcamp14/computervisioncompetition-cv-1#46]
## 📋 작업 개요
**작업 주제:** EfficientNet에서 LayoutLMv3로 전환 및 문서 분류 모델 최적화  
**담당자:** @username (실제 사용자명으로 대체 필요)  
**마감일:** 2025-09-15 24:00  
**예상 소요시간:** 총 6시간  

## 📅 기간
- 시작일: 2025-09-09
- 종료일: 2025-09-09 오후 2시

## 📌 목표
- EfficientNet-B3 기반 코드를 LayoutLMv3로 전환하여 문서 분류 성능 향상 (F1 스코어 0.95+ 목표).
- OCR 데이터 전처리 및 캐싱 구현으로 훈련 효율성 개선.
- K-Fold 앙상블과 TTA(Test-Time Augmentation)를 적용해 예측 정확도 극대화.

## ✅ Todo List
- [ ] OCR 데이터 전처리: `train.csv`와 `test.csv`에 EasyOCR로 추출한 텍스트와 바운딩 박스 추가 (`train_with_ocr.csv` 생성).
- [ ] LayoutLMv3Dataset 클래스 구현 및 DataLoader 최적화 (배치 크기=8, collate_fn 포함).
- [ ] LayoutLMv3ForSequenceClassification 모델로 훈련/검증 루프 수정 및 WandB 로깅 유지.
- [ ] TTA와 앙상블 적용: K-Fold 5개 모델과 5개 이미지 변형으로 최종 예측 생성.

## 💡 참고 사항
- OCR 처리 시 EasyOCR 사용 (GPU 필수, 신뢰도 0.5 이상 필터링).
- LayoutLMv3는 메모리 사용량이 높음. GPU 16GB 이상 추천, 배치 크기 조정 필요.
- 훈련 시간: K-Fold 5회, 에폭 10~20, 약 2~3배 느림 (EfficientNet 대비).
- Hugging Face LayoutLMv3 문서 참고: https://huggingface.co/microsoft/layoutlmv3-base

### 유용한 링크
- [ ] Kaggle 유사 솔루션: [Document Classification with LayoutLM](https://www.kaggle.com/code/example/layoutlm-example)  
- [ ] PyTorch 공식 예제: [Transformers for Sequence Classification](https://pytorch.org/hub/huggingface_pytorch-transformers/)  
- [ ] Hugging Face Transformers 문서: [LayoutLMv3 Documentation](https://huggingface.co/docs/transformers/model_doc/layoutlmv3)  

---

### 실험 중 발견사항
- LayoutLMv3는 텍스트와 레이아웃 정보를 활용해 F1 스코어가 기존 EfficientNet(0.9381)보다 향상될 가능성 높음 (0.95+ 목표).
- OCR 품질이 성능에 큰 영향. 손글씨나 저품질 이미지는 Google Cloud Vision API로 대체 검토.
- 배치 크기 8에서 메모리 안정적, 16 시 OutOfMemory 에러 발생 가능.

### 다음 실험 아이디어
- LayoutLMv3-large 모델 테스트 (더 높은 성능 기대, 메모리 24GB 필요).
- OCR 전처리 최적화: 미리 캐싱된 JSON 파일 사용으로 훈련 속도 개선.
- Optuna로 하이퍼파라미터 튜닝 (학습률, 배치 크기, 에폭 수).
- 데이터 증강: 텍스트 노이즈 추가 (오타 시뮬레이션)로 모델 견고성 강화.

### 팀원 피드백
-

**Comment by nimnusmik** [comment]
## 권장 결정
현재 상황에서는 LayoutLMv3를 포기하고 EfficientNet으로 전환하는 것이 현명

<img width="1192" height="135" alt="Image" src="https://github.com/user-attachments/assets/9d0b763a-8ffc-43dd-aa45-1d2023a54f13" />

OCR 품질이 너무 낮아 LayoutLMv3의 장점을 활용할 수 없음
순수 이미지 분류가 더 적합할 가능성이 높음
개발 시간과 리소스를 고려할 때 효율적

**Comment by nimnusmik** [comment]
#  tf_efficientnetv2_b3로 분류기 B 만들기 

# 1차 시도: 

```
model_name = 'tf_efficientnetv2_b3'  
img_size = 384
LR = 2e-4
EPOCHS = 15
BATCH_SIZE = 32
num_workers = 8
num_classes = 4
```
###  4-CLASS SUBSET K-FOLD CROSS VALIDATION RESULTS
Fold 1: 0.2322
Fold 2: 0.3321
Fold 3: 0.2545
Fold 4: 0.2247
Fold 5: 0.2249

Mean CV F1: 0.2537 ± 0.0407
Best single fold: 0.3321


# 2차시도 
- 작은 데이터에선 작은배치가 더 안정적
```
model_name = 'tf_efficientnetv2_b3'  
img_size = 384
LR = 1e-5 
EPOCHS = 15
BATCH_SIZE = 16 
num_workers = 8
num_classes = 4  
```

- 작은 데이터셋에서는 label smoothing이 해로울 수 있음
```
loss_fn = nn.CrossEntropyLoss()  # label_smoothing 제거
```
### 4-CLASS SUBSET K-FOLD CROSS VALIDATION RESULTS

Fold 1: 0.2195
Fold 2: 0.1916
Fold 3: 0.1683
Fold 4: 0.2521
Fold 5: 0.2193

Mean CV F1: 0.2102 ± 0.0284
Best single fold: 0.2521

--- 
# 3차시도 
>LR너무 줄였음 

```
model_name = 'tf_efficientnetv2_b3'  
img_size = 384
LR = 5e-5 
EPOCHS = 15
BATCH_SIZE = 32
num_workers = 8
num_classes = 4 
```
- 클래스 가중치 (14클래스가 너무 적어)

```
class_weights = torch.tensor([1.0, 1.0, 1.0, 2.0]).to(device)
    loss_fn = nn.CrossEntropyLoss(weight=class_weights)
```
### 4-CLASS SUBSET K-FOLD CROSS VALIDATION RESULTS

Fold 1: 0.2949
Fold 2: 0.2275
Fold 3: 0.1625
Fold 4: 0.3119
Fold 5: 0.2108

Mean CV F1: 0.2415 ± 0.0551
Best single fold: 0.3119

---- 
# 4차시도

```
model_name = 'tf_efficientnetv2_b3'  
img_size = 384
LR = 5e-5 
EPOCHS = 15
BATCH_SIZE = 4
num_workers = 8
num_classes = 4 
```
- 보수적인 스케일러 

```
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=5, gamma=0.5)
```
### 4-CLASS SUBSET K-FOLD CROSS VALIDATION RESULTS

Fold 1: 0.3126
Fold 2: 0.3303
Fold 3: 0.3415

Mean CV F1: 0.3281 ± 0.0119
Best single fold: 0.3415
> 이제 껏 제일 좋다


--- 
# 6차시도

> 이미지 사이즈 키워보기 
```
model_name = 'tf_efficientnetv2_b3'  
img_size = 512
LR = 5e-5 
EPOCHS = 15
BATCH_SIZE = 8
num_workers = 8
num_classes = 4 
```

### 4-CLASS SUBSET K-FOLD CROSS VALIDATION RESULTS
Fold 1: 0.3238
Fold 2: 0.3417
Fold 3: 0.2880

Mean CV F1: 0.3178 ± 0.0223
Best single fold: 0.3417

```
model_name = 'tf_efficientnetv2_b3'  
img_size = 384
LR = 5e-5 
EPOCHS = 15
BATCH_SIZE = 4
num_workers = 8
num_classes = 4 
```
- 다시 기존의 부드러운 스케일러 
```
scheduler = CosineAnnealingLR(optimizer, T_max=EPOCHS)
```
Fold 1: 0.3238
Fold 2: 0.3417
Fold 3: 0.2880

Mean CV F1: 0.3178 ± 0.0223
Best single fold: 0.3417

> 음 얘가 아주 살짝 더 나은...

**Comment by nimnusmik** [comment]
# convnext_base 로 바꿀까??? 에 대한 답변
현재 상황에서는 convnext_base로 바꾸는 것을 권장하지 않습니다.

## 이유 분석

**ConvNeXT Base의 문제점:**
- **훨씬 더 큰 모델**: EfficientNetV2-B3 대비 파라미터 수가 2-3배
- **Overfitting 위험 증가**: 현재도 validation이 불안정한데 더 큰 모델은 상황 악화
- **350개 샘플에는 과도**: 이런 작은 데이터셋에는 오버스펙

**현재 EfficientNetV2-B3의 장점:**
- 이미 0.34 F1로 괜찮은 성능 달성
- 작은 데이터셋에 적합한 크기
- 효율적인 아키텍처

## 대안 제안

**더 작은 모델 실험:**
```python
model_name = 'tf_efficientnetv2_b1'  # b3 → b1으로 감소
# 또는
model_name = 'tf_efficientnetv2_b0'  # 가장 작은 버전
```

**ConvNeXT를 쓰고 싶다면:**
```python
model_name = 'convnext_tiny_384_in22ft1k'  # base 대신 tiny
```

현재 0.34 F1을 안정화시키는 것이 우선입니다. CosineAnnealingLR 변경 후 결과를 보고, 여전히 개선이 필요하면 **더 작은 모델**을 시도하는 것이 바람직합니다.

--- 


# tf_efficientnetv2_b1
> 위에서 가장 좋았던 설정으로 더 작은 모델 시도 

```model_name = 'tf_efficientnetv2_b3'  
img_size = 512
LR = 5e-5 
EPOCHS = 15
BATCH_SIZE = 8
num_workers = 8
num_classes = 4 
```

### 4-CLASS SUBSET K-FOLD CROSS VALIDATION RESULTS

Fold 1: 0.2283
Fold 2: 0.3043
Fold 3: 0.2643

Mean CV F1: 0.2656 ± 0.0311
Best single fold: 0.3043



> 아 뭔데 장난하냐

**Comment by nimnusmik** [comment]
# convnext_tiny_384_in22ft1k 로 분류기 B 만들기
```
img_size = 512
LR = 5e-5 
EPOCHS = 15
BATCH_SIZE = 8
num_workers = 8
num_classes = 4 
```

Fold 1: 0.4161
Fold 2: 0.4081
Fold 3: 0.3846

Mean CV F1: 0.4029 ± 0.0133
Best single fold: 0.4161

> convnext가 좋네


# convnext_base_384_in22ft1k 로 분류기 B 만들기
> LR을 8e-5 로 향상
```
model_name = 'convnext_base_384_in22ft1k'  
img_size = 384
LR = 8e-5 
EPOCHS = 15
BATCH_SIZE = 8
num_workers = 8
num_classes = 4 
```

old 1: 0.4796
Fold 2: 0.4360
Fold 3: 0.4299

Mean CV F1: 0.4485 ± 0.0221
Best single fold: 0.4796

> 뭐야 나이스

```
# convnext_large_384_in22ft1k
model_name = 'convnext_large_384_in22ft1k'  
img_size = 384
LR = 8e-5 
EPOCHS = 15
BATCH_SIZE = 8
num_workers = 8
num_classes = 4 
```



Fold 1: 0.3594
Fold 2: 0.5157
Fold 3: 0.3271

Mean CV F1: 0.4008 ± 0.0823
Best single fold: 0.5157


---

**Comment by nimnusmik** [comment]
# swin_large_patch4_window12_384_in22ft1k
> ? 증강 약하게 줬는데도 성능 너무 안좋고 오래걸림

```

model_name = 'swin_large_patch4_window7_224'  
img_size = 224
LR = 3e-5
EPOCHS = 15
BATCH_SIZE = 2
num_workers = 8
num_classes = 4 
```

**Comment by nimnusmik** [comment]
## 강사님 1:1 멘토링

<img width="782" height="736" alt="Image" src="https://github.com/user-attachments/assets/be93c504-1a51-4163-a0d0-141644d53fca" />

- [x] 여러 모델 시험해보기 
- [x]  confusion matrix 확인해 본후 focal loss 로 추가 보정 
confusion matrix에서 보면 class 2가 가장 성능이 안좋은거 같은데, 여기에서의 문제점 같은게 있을까 싶네요? 2->0으로 가장 많이 예측되는거 같은데..
```
class_weights = torch.tensor([1.0, 1.0, 4.0, 3.0]).to(device)
focal_loss = FocalLoss(alpha=[1.0, 1.0, 3.0, 4.0], gamma=2.0)  # gamma 2.0이 일반적

loss_fn = focal_loss
``` 

Fold 1: 0.2949
Fold 2: 0.1755
Fold 3: 0.2212

Mean CV F1: 0.2305 ± 0.0492
Best single fold: 0.2949

> 성능은 훨 별로...

- [ ] B 의성능이 안좋아도 붙여보면 다를 수 있음 A에 붙이는거 도입후 비교해보기
